presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java

```diff
@@ -0,0 +1,195 @@
+import java.util.List;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class AssignUniqueIdOperator
+        implements Operator
+{
+    private static final long VALUES_PER_REQUEST = 1 << 20;
+
+    public static class AssignUniqueIdOperatorFactory
+            implements OperatorFactory
+    {
+        private final int operatorId;
+        private final PlanNodeId planNodeId;
+        private final List<Type> types;
+        private boolean closed;
+        private final AtomicLong valuePool = new AtomicLong();
+
+        public AssignUniqueIdOperatorFactory(
+                int operatorId,
+                PlanNodeId planNodeId,
+                List<? extends Type> types)
+        {
+            this.operatorId = operatorId;
+            this.planNodeId = requireNonNull(planNodeId, "planNodeId is null");
+            this.types = ImmutableList.copyOf(types);
+        }
+
+        @Override
+        public List<Type> getTypes()
+        {
+            return types;
+        }
+
+        @Override
+        public Operator createOperator(DriverContext driverContext)
+        {
+            checkState(!closed, "Factory is already closed");
+
+            OperatorContext operatorContext = driverContext.addOperatorContext(
+                    operatorId,
+                    planNodeId,
+                    AssignUniqueIdOperator.class.getSimpleName());
+            return new AssignUniqueIdOperator(operatorContext, types, valuePool);
+        }
+
+        @Override
+        public void close()
+        {
+            closed = true;
+        }
+
+        @Override
+        public OperatorFactory duplicate()
+        {
+            return new AssignUniqueIdOperatorFactory(operatorId, planNodeId, types);
+        }
+    }
+
+    private final OperatorContext operatorContext;
+    private boolean finishing;
+    private final AtomicLong valuePool;
+    private final List<Type> types;
+    private final LongValueCompressor valueCompressor = new LongValueCompressor();
+
+    private Page inputPage;
+    private long counter;
+    private long lastCounterValue;
+
+    public AssignUniqueIdOperator(
+            OperatorContext operatorContext,
+            List<Type> types,
+            AtomicLong valuePool)
+    {
+        this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
+        this.types = ImmutableList.copyOf(types);
+        this.valuePool = requireNonNull(valuePool, "valuePool is null");
+
+        TaskId fullTaskId = operatorContext.getDriverContext().getTaskId();
+
+        valueCompressor.store(fullTaskId.getStageId().getId(), 10)
+                .store(fullTaskId.getId(), 14);
+
+        requestValues();
+    }
+
+    private void requestValues()
+    {
+        lastCounterValue = valuePool.addAndGet(VALUES_PER_REQUEST);
```

Comment:
Also, lastCounterValue is misnamed, since the current batch will not generate a row with that value. Maybe "maxValue"?